src/main/scala/io/dronekit/cloud/SignRequestForAWS.scala

```diff
@@ -0,0 +1,302 @@
+    Hex.encodeHexString(HmacSHA256(stringToSign, signatureKey))
+  }
+
+  def createAuthorizationHeader(httpRequest: HttpRequest, key: String, region: String, accessKeyId: String, service: String): Future[String] = {
+    val canonicalRequestFuture: Future[String] = createCanonicalRequest(httpRequest)
+    canonicalRequestFuture.map{canonicalRequest =>
+      val stringToSign = createStringToSign(httpRequest, canonicalRequest,region, service)
+      val signatureKey = createSignatureKey(httpRequest.headers, httpRequest.uri, key, region, service)
+      val signature = getSignature(signatureKey, stringToSign)
+      val credential = accessKeyId + "/" + getCredentialScope(httpRequest.headers, httpRequest.uri,region, service)
+      val signedHeaders = getSignedHeaders(httpRequest.headers, httpRequest.uri)
+      val algorithm = "AWS4-HMAC-SHA256"
+      algorithm + " Credential=" + credential + ", SignedHeaders=" + signedHeaders + ", Signature=" + signature
+    }
+  }
+
+  // adds the authorizations header to an httpRequest as described in http://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html
+  def addAuthorizationHeader(httpRequest: HttpRequest, key: String, region: String, accessKeyId: String, service: String): Future[HttpRequest] = {
+    val request = HttpRequest(
+      method = httpRequest.method,
+//      uri = createUri(httpRequest.uri),
+      uri = httpRequest.uri,
+      entity = httpRequest.entity,
+      headers = httpRequest.headers :+ RawHeader("x-amz-date", getUTCTime()),
+      protocol = httpRequest.protocol
+    )
+    val authHeaderFuture = createAuthorizationHeader(request, key, region, accessKeyId, service)
+    authHeaderFuture.map { authHeader => request.withHeaders(request.headers :+ RawHeader("Authorization", authHeader))
+    }
+  }
+
+  // adds the authorization query to the uri for aws services as described in http://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html
+  def addQueryString(httpRequest: HttpRequest, key: String, region: String, accessKeyId: String, service: String, expires: Int): Future[HttpRequest] = {
+    val date = getUTCTime()
+    val dateHeader = RawHeader("x-amz-date", date)
+    val algorithm = "AWS4-HMAC-SHA256"
+    val credential = accessKeyId + "/" + getCredentialScope(httpRequest.headers :+ dateHeader, httpRequest.uri,region, service)
+    val signedHeaders = getSignedHeaders(httpRequest.headers :+ dateHeader, httpRequest.uri)
+    var requestString = httpRequest.uri.query.toString()
+    if (requestString.length > 1) {
+      requestString = requestString + "&"
+    }
+    val qString = uriEncode(requestString + "X-Amz-Algorithm=" + algorithm + "&X-Amz-Credential=" + credential +
+                  "&X-Amz-Date=" + date + "&X-Amz-Expires=" + expires.toString + "&X-Amz-SignedHeaders=" + signedHeaders)
+    val URI = Uri(httpRequest.uri.scheme, httpRequest.uri.authority,
+        Path(httpRequest.uri.path.toString().replace("+","%20")),
```

Comment:
uriEncode should be doing this for you, you should not be manually replacing parameters.